lib/task-queue/TaskRouter.js

```diff
@@ -0,0 +1,133 @@
+/* globals Task */
+"use strict";
+
+const {Cu} = require("chrome");
+const {WorkerTask} = require("lib/task-queue/WorkerTask");
+const {Storage} = require("lib/task-queue/Storage");
+const {Page} = require("lib/task-queue/Page");
+
+Cu.import("resource://gre/modules/Task.jsm");
+
+// The lock for page and task creation test-and-set
+let taskCreationLock = false;
+// The queue for messages waiting to be processed
+const mutexQueue = [];
+
+class TaskRouter {
+  /**
+   * Creates a new task router from a map of messages to TaskProcessor arrays
+   *
+   * @param {Map<string, TaskProcessor[]>} processorMap
+   *                                       A map of Task Processors arrays for each message
+   */
+  constructor(processorMap) {
+    if (!processorMap) {
+      throw new Error("TaskRouter needs a map as argument.");
+    }
+    // Initialize task processors
+    this.routes = processorMap;
+  }
+
+  /**
+   * Check if the page should be processed by the queue
+   *
+   * @param {String}    url
+   *                    The URL of the page we want to check
+   * @returns {Promise}
+   *                    Returns true if the page should be processed, false otherwise
+   */
+  asyncNeedProcessing(url) {
+    return Task.spawn((function*() {
+      // Do we already have a task for this url?
+      const tasks = yield WorkerTask.asyncGetByUrl(url);
```

Comment:
given that this query runs for every page and that we don't need the data, we should have a function in `Task` that returns a boolean. Perhaps we can call this function `asyncURLExists`.

The SQL query could be:

```sql
SELECT 1 FROM moz_tasks WHERE url = %s
```